### PR TITLE
Show visibility status of email in own profile

### PR DESCRIPTION
**Pull request by n0toose**
I've heard many reports of users getting scared when they see their own
email address for their own profile, as they believe that the email
field is also visible to other users. Currently, using Incognito mode
or going over the Settings is the only "reasonable" way to verify this
from the perspective of the user.

A locked padlock should be enough to indicate that the email is not
visible to anyone apart from the user and the admins. An unlocked
padlock is used if the email address is only shown to authenticated
users.

Some additional string-related changes in the Settings were introduced
as well to ensure consistency, and the comments in the relevant tests
were improved so as to allow for easier modifications in the future.

---

#### Screenshot (EDIT: Scroll down for more up-to-date screenshots)

***Please remove this section before merging.***

image

This lock should only appear if the email address is explicitly hidden using the `Hide Email Address` setting. The change was originally tested on top of and designed for the Forgejo fork, but I don't expect any problems to arise from this and I don't think that a documentation-related change is strictly necessary.

### CODE REVIEW

**Comment by n0toose** [comment]
I just made a force-push as I made a typo in the commit summary, apologies :D

**Comment by n0toose** [comment]
This change was apparently made redundant last week and I didn't precisely notice that change, apologies once again: https://github.com/go-gitea/gitea/commit/6706ac2a0f5f2fe4f8e2555be7e2a8b4d5946398

I'll revamp the PR as I believe that I could improve the UX a little bit with my own idea.

**Comment by n0toose** [comment]
Here's a new concept:

image

image

**Comment by n0toose** [comment]
Hi, sorry for the mess. I introduced the "show main email on the user's profile" thing again, but ended up using icons to clearly differentiate between public and no visibility. I think this is consistent with the overall look of the personalized Profile page, as similar indicators are present next to the users' repositories, and this leaves some room for additional, purely hypothetical, changes in the future, like e.g. "only users that are logged in can see my email".

I am not used to modifying templates, this is a first contribution and I see that the coverage has dropped. Feedback is appreciated.

**Comment by silverwind** [comment]
Hmm, instead of adding an additional icon which does not really fit there, why not indicate on mouseover via `data-tooltip-content`?

**Comment by n0toose** [comment]
> why not indicate on mouseover via `data-tooltip-content`

With all due respect, I don't think that non-obvious xkcd-style "hover over it" tips will be helpful when you e.g. think that your email address has been leaked. I think that not including the address at all, which is the case right now, would be relatively better.

**Comment by n0toose** [comment]
oh, uh, this seems to be breaking the tests now :D

**Comment by n0toose** [comment]
> IMHO, add right-side icon https://primer.github.io/octicons/info-16, with data-tooltip-content via translation "Only logged-in users can see your e-mail address".

I don't want to get super annoying about this, but I am not sure that not using clear iconography to describe the privacy status is productive. I can see this making a lot of sense with the icons that I am using right now, but I don't believe that an ℹ️ box with a hint should not adequately fix the "super scary and confusing UI" problem that I brought up - I'd like for a hint to play a complementary role to confirm what the user is seeing, instead of showing them something in such a confusing way that prompted https://github.com/go-gitea/gitea/commit/6706ac2a0f5f2fe4f8e2555be7e2a8b4d5946398 (aka. just the email address, with the addition of an omnipresent info box) and then telling them "oh, you're fine, everything is the way you want it" in a small box.

**Comment by n0toose** [comment]
> Also check if there are other conditions besides the logged-in condition. There might be an option to always hide the address from everyone but yourselves.

Oh, sorry, just noticed that I kind of messed up that part after I tried to scramble everything together to adjust my change so that it'd work with this codebase.

I did it the way I wanted to do it, but I also used hints and also a hyperlink that leads the user directly to the right checkbox.

### Logged in

image

image

### Logged out

(Email address is not supposed to be hidden here.)

image

**Comment by n0toose** [comment]
Hi again, thanks for your patience. The PR isn't ready yet, as for some reason, [.ShowUserEmail](https://github.com/go-gitea/gitea/pull/23900/files#diff-45ac589738914062c84c666ad32b2fd77e8d4a0bc485aefd8f412ece7c16b100R50-R55) doesn't work as it should anymore.

Accessing the user profile in Incognito mode when **.ShowUserEmail** is removed *always* shows the email address without the icons (so, the first condition fails, else-block is accessed properly) but re-adding it again means that the email address will *never* be shown, regardless of the setting that the user has set. I am not sure why this would break, as this is how this basically worked before. I'd very much appreciate any assistance and any further feedback :D

**Comment by n0toose** [comment]
Yeah, sorry for this mess, I just kept thinking that this PR was the way I wanted it to be and then kept incrementally working on more and more details that I kept noticing along the way.

image

I replaced the globe with an "unlocked emoji", following the example of other applications like Mastodon ("Unlisted" uses the same icon and it stands for "making a public post that doesn't appear in public timelines", aka. "public, but not exactly quite") thus leaving the possibility for a third, hypothetical option that will show an email address to everyone on the Internet open, without having to change the meaning of the icons at a later stage.

I made the icon right-aligned to tone down the "weird" factor.

I stopped using weird variables that I wasn't supposed to be using and I started using the variables that I was supposed to use in the first place.

I improved the consistency of the tool tips, and made one out-of-scope change in the user settings because it somewhat bothered me.

**Comment by fogti** [comment]
ok, seems this obviously breaks the test case which checks that the user email shouldn't be visible at all...
>         	Test:       	TestSettingShowUserEmailProfile

**Comment by n0toose** [comment]
> ok, seems this obviously breaks the test case which checks that the user email shouldn't be visible at all...

Thanks! Even I wouldn't find it very fun for this PR to get stalled because of me not having attended to this yet, I don't have the capacity to figure this out right now. I'll see.

**Comment by fogti** [comment]
That unit test (and possibly related ones idk) seem to be defined in https://github.com/go-gitea/gitea/blob/376396a08838031bb351c9ed130baf7264adb72f/tests/integration/setting_test.go#L44

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
keep_email_private_popup = Your email address will only be visible to you and the administrators
```

**Comment by silverwind** [comment]
I like this new display. Translation changes are fine as well, except that one suggestion above. Once you figure out the tests, this is LGTM.

**Review by lunny — APPROVED** [review]

**Comment by silverwind** [comment]
Current CI failure is unrelated, we need to update test-env to go 1.20.3.

**Comment by n0toose** [comment]
> Once you figure out the tests, this is LGTM

Don't have enough time to work on these as I'm currently in transit, but I am inclined to believe that the problem is that *another* user account should be checking for the same conditions.

**Comment by silverwind** [comment]
Still need 1 approval, and tests are failing.

**Comment by fogti** [comment]
so, I'd guess in the file linked above lines 79 and/or 88 need to be changed (e.g. `s/NotContains/Contains/` on these lines). The tests appear to be failing for unrelated reasons (security advisory scanning, but this PR so far hasn't modified dependencies or actual non-template code, so I'd assume that it's unrelated), which are perpendicular to this PR.

**Comment by n0toose** [comment]
I think https://github.com/go-gitea/gitea/blob/376396a08838031bb351c9ed130baf7264adb72f/tests/integration/setting_test.go#L64-L65 and https://github.com/go-gitea/gitea/blob/376396a08838031bb351c9ed130baf7264adb72f/tests/integration/setting_test.go#L74-L90 have to be removed from the test cases without introducing anything new, as https://github.com/go-gitea/gitea/blob/376396a08838031bb351c9ed130baf7264adb72f/tests/integration/setting_test.go#L53-L58 checks whether the email is still present in the user profile.

Those lines should be removed and no additional checks should be introduced, if I am not wrong. I will take another careful look later, just wanted to post a small status update.

**Comment by n0toose** [comment]
Nevermind, I just changed the last part (`// user2 can not see self`) and improved the comments a little bit.

**Comment by n0toose** [comment]
(p.s. before squashing and merging, please check that the attribution using my full name is correct and that github isn't adding some `Co-authored-by` tag using my username instead of my full name, as the merge commit used my username instead, i'd rebase it all myself, but that would delete the previous review and potentially cause more trouble, ty ❤️)

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
email_visibility.private = Your email address is only visible to you and administrators
```

**Review by silverwind — APPROVED** [review]

**Comment by n0toose** [comment]
Maybe this string should be altered as well: https://github.com/go-gitea/gitea/blob/81174623d134315a97193b432da48971e62f8379/options/locale/locale_en-US.ini#L615

**Comment by silverwind** [comment]
Yeah but that is not related to this PR, would do in another PR.

**Comment by n0toose** [comment]
Got it.

**Comment by silverwind** [comment]
> Co-authored-by



> (p.s. before squashing and merging, please check that the attribution using my full name is correct and that github isn't adding some `Co-authored-by` tag using my username instead of my full name, as the merge commit used my username instead, i'd rebase it all myself, but that would delete the previous review and potentially cause more trouble, ty ❤️)

Checked the merge message, there is only a `Co-authored-by` for me, you are the author.

**Comment by n0toose** [comment]
> Checked the merge message, there is only a `Co-authored-by` for me, you are the author.

if it contains the username `n0toose` and not full name as it is visible in the `git log -p`, I'd appreciate a correction :D

**Comment by silverwind** [comment]
> if it contains the username `n0toose` and not full name as it is visible in the `git log -p`, I'd appreciate a correction :D

It contains your original username from the initial commit. I can't alter this on the UI here, and also would prefer not to :)

**Comment by n0toose** [comment]
Should be fine then, I'm just conditioned to double-checking that my information does not show up double / "two authors, one email" as e.g. Gitea does not handle that as well as GitHub and that ends up ruining things like internal contribution statistics. Sorry and thanks, I am just a bit of a perfectionist even if it doesn't matter in the end of the day :)

**Pull request merged**
